pkg/util/status/status.go

```diff
@@ -0,0 +1,300 @@
+		return u.patchWithSubresource(obj, pt, data)
+	} else {
+		return u.patchWithoutSubresource(obj, pt, data)
+	}
+}
+
+// updateWithoutSubresource will try to update the  status via PUT sent to the main REST endpoint.
+// If status of the returned object did not change, it knows that it should have used the /status subresource
+// and will switch the updater itself over to permanently use the /status subresource.
+func (u *updater) updateWithoutSubresource(obj runtime.Object) (err error) {
+	oldStatus, newStatus, err := u.updateUnstructured(obj)
+	if err != nil {
+		return err
+	}
+	if !reflect.DeepEqual(oldStatus, newStatus) {
+		u.setSubresource(true)
+		return u.updateStatusUnstructured(obj)
+	}
+	return nil
+}
+
+// updateWithSubresource will try to update the  status via PUT sent to the /status subresource.
+// If a 404 error is returned, it will try the main rest entrypoint instead. In case that this
+// call succeeds, it will switch the updater to permanently use the main entrypoint.
+func (u *updater) updateWithSubresource(obj runtime.Object) (err error) {
+	err = u.updateStatusUnstructured(obj)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if errors.IsNotFound(err) {
+		oldStatus, newStatus, err := u.updateUnstructured(obj)
+		if errors.IsNotFound(err) {
+			// object does not exist
+			return err
+		}
+		if err == nil && reflect.DeepEqual(oldStatus, newStatus) {
+			u.setSubresource(false)
+		}
+		return err
+	} else {
+		return err
+	}
+}
+
+// patchWithoutSubresource will try to update the  status via PATCH sent to the main REST endpoint.
+// If the resource version of the returned object did not change, it knows that it should have used the /status subresource
+// and will switch the updater itself over to permanently use the /status subresource.
+func (u *updater) patchWithoutSubresource(obj runtime.Object, patchType types.PatchType, data []byte) (err error) {
+	oldResourceVersion, newResourceVersion, err := u.patchUnstructured(obj, patchType, data)
+	if err != nil {
+		return err
+	}
+	if oldResourceVersion == newResourceVersion {
+		u.setSubresource(true)
+		return u.patchStatusUnstructured(obj, patchType, data)
+	}
+	return nil
+}
+
+// patchWithSubresource will try to update the  status via PATCH sent to the /status subresource.
+// If a 404 error is returned, it will try the main rest entrypoint instead. In case that this
+// call succeeds, it will switch the updater to permanently use the main entrypoint.
+func (u *updater) patchWithSubresource(obj runtime.Object, patchType types.PatchType, data []byte) (err error) {
+	err = u.patchStatusUnstructured(obj, patchType, data)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if errors.IsNotFound(err) {
+		oldResourceVersion, newResourceVersions, err := u.patchUnstructured(obj, patchType, data)
+		if errors.IsNotFound(err) {
+			// object does not exist
+			return err
+		}
+		if err == nil && oldResourceVersion != newResourceVersions {
```

Comment:
like update, we need an error if old == new here.

Reply:
The same like above. If there was no error and the subresource changed we have to permanently switch from PatchStatus to Patch. In any other case we just return whatever err is (nil or any other general error).

Reply:
You were right. Addressed and added a test.